internal/lefthook/run.go

```diff
@@ -10,6 +49,313 @@ func Run(opts *Options, hookName string, gitArgs []string) error {
+		return errors.New("hook is not found")
+	}
+	if err := hook.Validate(); err != nil {
+		log.Info(log.Red(fmt.Sprintf("Config error! %s", err)))
+		return err
+	}
+
+	startTime := time.Now()
+
+	if !isOutputDisabled(outputMeta, cfg.SkipOutput) {
+		log.Info(log.Cyan("Lefthook v" + version.Version()))
+		log.Info(log.Cyan("RUNNING HOOK:"), log.Bold(hookName))
+	}
+
+	spinner.Start()
+
+	var wg sync.WaitGroup
+
+	l.executeScripts(filepath.Join(cfg.SourceDir, hookName), gitArgs, hook, &wg)
+	l.executeScripts(filepath.Join(cfg.SourceDirLocal, hookName), gitArgs, hook, &wg)
+
+	for _, name := range hook.CommandsSorted {
+		wg.Add(1)
+
+		if hook.Parallel {
+			go l.executeCommand(name, gitArgs, hook, &wg)
+		} else {
+			l.executeCommand(name, gitArgs, hook, &wg)
+		}
+	}
+
+	wg.Wait()
+	spinner.Stop()
+
+	if !isOutputDisabled(outputSummary, cfg.SkipOutput) {
+		printSummary(time.Since(startTime))
+	}
+
+	if len(failList) == 0 {
+		return nil
+	}
+	return errors.New("Have failed script")
+}
+
+func (l *Lefthook) executeCommand(commandName string, gitArgs []string, hook *config.Hook, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if hook.Piped && isPipeBroken {
+		spinner.RestartWithMsg("\n", log.Bold(commandName), log.Red("(SKIP BY BROKEN PIPE)"))
+		return
+	}
+
+	command := hook.Commands[commandName]
+
+	if err := command.Validate(); err != nil {
+		appendFailList(commandName)
+		setPipeBroken()
+		spinner.RestartWithMsg(commandName, log.Red(fmt.Sprintf("(SKIP BY CONFIG ERROR: %s)", err)))
+		return
+	}
+	if command.DoSkip(l.repo.State()) {
+		spinner.RestartWithMsg(successSprint(log.Bold(commandName), log.Red("(SKIP BY SETTINGS)")))
+		return
+	}
+	if isContainsExcludedTags(command.Tags, hook.ExcludeTags) {
+		spinner.RestartWithMsg(successSprint(log.Bold(commandName), log.Red("(SKIP BY TAGS)")))
+		return
+	}
+
+	filesCommand := hook.Files
+	if command.Files != "" {
+		filesCommand = command.Files
+	}
+
+	runner := command.Run
+	if strings.Contains(runner, config.SubStagedFiles) {
+		files, err := l.repo.StagedFiles()
+		files = prepareFiles(command, files)
+		if err == nil {
+			runner = strings.ReplaceAll(runner, config.SubStagedFiles, strings.Join(files, " "))
+		}
+	}
+	if strings.Contains(runner, config.SubFiles) || filesCommand != "" {
+		files, err := git.FilesByCommand(filesCommand)
+		files = prepareFiles(command, files)
+		if err == nil {
+			runner = strings.ReplaceAll(runner, config.SubFiles, strings.Join(files, " "))
+		}
+	}
+	if strings.Contains(runner, config.PushFiles) {
+		files, err := l.repo.PushFiles()
+		files = prepareFiles(command, files)
+		if err == nil {
+			runner = strings.ReplaceAll(runner, config.PushFiles, strings.Join(files, " "))
+		}
+	}
+	if strings.Contains(runner, config.SubAllFiles) {
+		files, err := l.repo.AllFiles()
+		files = prepareFiles(command, files)
+		if err == nil {
+			runner = strings.ReplaceAll(runner, config.SubAllFiles, strings.Join(files, " "))
+		}
+	}
```

Comment:
Can we DRY this?

Reply:
Do you have any suggestions? I can't find any obvious solutions.

Reply:
What about creating an array of anonymous structures, that hold the dependencies of each part? (the substitution template and a git function that returns the files and error)

Reply:
Yeah, I thought about that, but one of the functions have a different signature.

Reply:
I think about something like this:

```go
func replaceTemplate(runner, sub string, f func()([]string, error)) string {
  if !strings.Contains(runner, config.SubStagedFiles) {
    return runner
  }
  
  files, err := f()
  // ...
}
```

Reply:
I'll just wrap it